src/msg/async/rdma/RDMAStack.cc

```diff
@@ -128,36 +128,60 @@ void RDMADispatcher::handle_async_event()
       return;
     }
     perf_logger->inc(l_msgr_rdma_total_async_events);
+    ldout(cct, 1) << __func__ << "Event : " << ibv_event_type_str(async_event.event_type) << dendl;
+
     switch (async_event.event_type) {
       /***********************CQ events********************/
       case IBV_EVENT_CQ_ERR:
-        lderr(cct) << __func__ << " CQ Overflow, dev = " << ib->get_device()->ctxt
+        lderr(cct) << __func__ << " Fatal Error, effect all QP bound with same CQ, "
+                   << " CQ Overflow, dev = " << ib->get_device()->ctxt
                    << " Need destroy and recreate resource " << dendl;
         break;
       /***********************QP events********************/
       case IBV_EVENT_QP_FATAL:
-        /* Error occurred on a QP and it transitioned to error state */
-        lderr(cct) << __func__ << " Error occurred on a QP and it transitioned to error state, dev = "
-                   << ib->get_device()->ctxt << " Need destroy and recreate resource " << dendl;
+        {
+          /* Error occurred on a QP and it transitioned to error state */
+          ibv_qp* ib_qp = async_event.element.qp;
+          uint32_t qpn = ib_qp->qp_num;
+          std::lock_guard l{lock};
+          QueuePair* qp = get_qp_lockless(qpn);
+          lderr(cct) << __func__ << " Fatal Error, event associate qp number: " << qpn
+                     << " Queue Pair status: " << Infiniband::qp_state_string(qp->get_state())
+                     << " Event : " << ibv_event_type_str(async_event.event_type) << dendl;
+        }
         break;
       case IBV_EVENT_QP_LAST_WQE_REACHED:
-        /* Last WQE Reached on a QP associated with and SRQ */
         {
-          // FIXME: Currently we must ensure no other factor make QP in ERROR state,
-          // otherwise this qp can't be deleted in current cleanup flow.
+          /*
+           * 1. The QP bound with SRQ is in IBV_QPS_ERR state & no more WQE on the RQ of the QP
+           *    Reason: QP is force switched into Error before posting Beacon WR.
```

Comment:
Why not to post a second beacon to the RQ in case of !SRQ? Otherwise you mix both drain semantics, which leads to bugs.  And this LAST_WQE_REACHED event is not supported on qlogic.

Reply:
@rouming 
I think there's no need to post another Beacon to RQ:
1. dead_queue_pairs will be reclaimed (call QueuePair destruction) only under there's no CQE on completion queue.
2. When QueuePair is switched into Error state, all outstanding WRs will be flushed into CQ, which means that WRs on RQ will be flushed into CQ. So, only after all WR-CQEs are polled from completion channel, then the corresponding QueuePair could be reclaimed.

Reply:
We have two completions queues and when SRQ is replaced with RQ I would drain it as well, otherwise I do not see how two different CQs paired with RQ and SQ can be synchronized between each other. Again, take a look how it is in kernel, say mlx5 driver is implemented: https://elixir.bootlin.com/linux/latest/source/drivers/infiniband/hw/mlx5/qp.c#L6344 and the caller https://elixir.bootlin.com/linux/latest/source/drivers/infiniband/core/verbs.c#L2660, where RQ is drained as well if SRQ is off.

I still do not understand why you keep async LAST event.

Reply:
@rouming 
We have separate CQ for SR/RQ. 
>otherwise I do not see how two different CQs paired with RQ and SQ can be synchronized between each other.
It's here: **if (!tx_ret && !rx_ret)**
After switching the QP into error state, outstanding WRs on RQ will be flushed into its own CQ. 
One rx_ret is zero it means that all the WRs on RQ have been flushed into CQ and been retrieved from CQ. If there're outstanding WRs on RQ when QP is in error state, rx_ret can't be zero

Reply:
@rouming Do you still have concern if we don't post Beacon WR to RQ?

Reply:
> @rouming Do you still have concern if we don't post Beacon WR to RQ?

For RQ case (not SRQ)? This should be done using post_recv, exactly like for WR case, where post_send is used.

Reply:
@rouming 
I'll mark this conversation as being closed.
As we've talked in email, there's no need to post another WR to RQ(no SRQ).